### PR TITLE
feat: dual-clip in grpo loss

### PR DESCRIPTION
**Pull request by ZhiyuLi-Nvidia**
# What does this PR do ?

Add dual-clip support in grpo loss

# Issues
List issues that this PR closes ([syntax](https://docs.github.com/en/issues/tracking-your-work-with-issues/using-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)):

# Usage
* **You can potentially add a usage example below**

```python
# Add a code snippet demonstrating how to use this 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](/NVIDIA/nemo-rl/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [x] Did you run the unit tests and functional tests locally? Visit our [Testing Guide](/NVIDIA/nemo-rl/blob/main/docs/testing.md) for how to run tests
- [x] Did you add or update any necessary documentation? Visit our [Document Development Guide](/NVIDIA/nemo-rl/blob/main/docs/documentation.md) for how to write, build and test the docs.

# Additional Information
* ...

### CODE REVIEW

File: nemo_rl/algorithms/loss_functions.py [hunk 1]
```diff
@@ -73,12 +73,22 @@ class ClippedPGLossFn(LossFunction):
     For REINFORCE/RLOO (when disable_ppo_ratio=True), the formula simplifies to:
     L(θ) = E_t [ π_θ(a_t|s_t) * A_t ] - β * KL(π_θ || π_ref)
 
+    According to the Dual-clipping paper (https://arxiv.org/pdf/1912.09729),
```
Comment (SahilJain314) [inline nemo_rl/algorithms/loss_functions.py]:
```suggestion
    Also supports "Dual-Clipping" from https://arxiv.org/pdf/1912.09729, which
```

File: nemo_rl/algorithms/loss_functions.py [hunk 2]
```diff
@@ -73,12 +73,22 @@ class ClippedPGLossFn(LossFunction):
...
+    the paper imposes an additional upper bound on the probability ratio, where curr_lp >> prev_lp
```
Comment (SahilJain314) [inline nemo_rl/algorithms/loss_functions.py]:
```suggestion
    imposes an additional upper bound on the probability ratio when advantages are negative.
```

File: nemo_rl/algorithms/loss_functions.py [hunk 3]
```diff
@@ -73,12 +73,22 @@ class ClippedPGLossFn(LossFunction):
...
+    to prevent excessive policy updates when dealing with strongly negative advantages,
```
Comment (SahilJain314) [inline nemo_rl/algorithms/loss_functions.py]:
```suggestion
    This prevents excessive policy updates. $rA << 0$ -> $cA$(clipped) 
```

File: nemo_rl/algorithms/loss_functions.py [hunk 4]
```diff
@@ -73,12 +73,22 @@ class ClippedPGLossFn(LossFunction):
...
+    the loss function is modified to the following when A_t < 0:
```
Comment (SahilJain314) [inline nemo_rl/algorithms/loss_functions.py]:
```suggestion
    The loss function is modified to the following when A_t < 0:
```

File: nemo_rl/algorithms/loss_functions.py [hunk 5]
```diff
@@ -73,12 +73,22 @@ class ClippedPGLossFn(LossFunction):
...
+    L(θ) = E_t [ max(min(r_t(θ) * A_t, clip(r_t(θ), 1-ε, 1+ε) * A_t), c * A_t) ] - β * KL(π_θ || π_ref)
+
+    where:
+    - c is the dual-clip parameter (ratio_clip_c) and must be greater than 1 and set as 3 empirically
```
Comment (SahilJain314) [inline nemo_rl/algorithms/loss_functions.py]:
```suggestion
    - c is the dual-clip parameter (ratio_clip_c), which must be greater than 1 and is
      usually set as 3 empirically.
```

File: nemo_rl/algorithms/loss_functions.py [hunk 6]
```diff
@@ -162,6 +173,19 @@ def __call__(
         loss1 = -advantages * ratios
         loss2 = -advantages * ratios_clamped
 
+        # Determine which value to use for clipping (max for pessimistic estimate)
+        clip_loss = torch.max(loss1, loss2)
+
+        # Dual-clipping see https://arxiv.org/pdf/1912.09729
+        if self.ratio_clip_c is not None:
+            assert self.ratio_clip_c > 1, (
+                "ratio_clip_c must exceed 1 representing a lower bound of the ratios"
```
Comment (parthchadha) [inline nemo_rl/algorithms/loss_functions.py]:
```suggestion
                f"ratio_clip_c must exceed 1 representing a lower bound of the ratios, got {self.ratio_clip_c}."
```

File: nemo_rl/algorithms/loss_functions.py [hunk 7]
```diff
@@ -73,12 +73,23 @@ class ClippedPGLossFn(LossFunction):
     For REINFORCE/RLOO (when disable_ppo_ratio=True), the formula simplifies to:
     L(θ) = E_t [ π_θ(a_t|s_t) * A_t ] - β * KL(π_θ || π_ref)
 
+    Also supports "Dual-Clipping" from https://arxiv.org/pdf/1912.09729, which
```
Comment (parthchadha) [inline nemo_rl/algorithms/loss_functions.py:77]: We should update `grpo.md` as well.

File: nemo_rl/algorithms/loss_functions.py (hunk 6 above)
Comment (ZhiyuLi-Nvidia) [reply nemo_rl/algorithms/loss_functions.py]: Thank you @parthchadha

File: nemo_rl/algorithms/loss_functions.py [hunk 8]
```diff
@@ -73,12 +73,23 @@ class ClippedPGLossFn(LossFunction):
...
+    imposes an additional upper bound on the probability ratio when advantages are negative.
+    This prevents excessive policy updates. $rA << 0$ -> $cA$(clipped)
```
Comment (SahilJain314) [inline nemo_rl/algorithms/loss_functions.py]:
```suggestion
    This prevents excessive policy updates. $rA << 0$ -> $cA$(clipped)
```

Comment (ZhiyuLi-Nvidia) [reply nemo_rl/algorithms/loss_functions.py]: Thank you @SahilJain314

File: nemo_rl/algorithms/loss_functions.py (hunk 7 above)
Comment (ZhiyuLi-Nvidia) [reply nemo_rl/algorithms/loss_functions.py:77]: Done.

image

File: nemo_rl/algorithms/loss_functions.py [hunk 9]
```diff
@@ -73,12 +73,23 @@ class ClippedPGLossFn(LossFunction):
...
+    - c is the dual-clip parameter (ratio_clip_c), which must be greater than 1 and is
+      usually set as 3 empirically.
+
     Due to potential numerical instability, we cast the logits to float32 before computing the loss.
     """
 
     def __init__(self, cfg: ClippedPGLossConfig):
         self.ratio_clip_min = cfg["ratio_clip_min"]
         self.ratio_clip_max = cfg["ratio_clip_max"]
+        self.ratio_clip_c = cfg.get("ratio_clip_c", None)
```
Comment (terrykong) [inline nemo_rl/algorithms/loss_functions.py]: can you use bracket notation to avoid defaults being nested deeply in our code? (we should error loudly)

Comment (ZhiyuLi-Nvidia) [reply nemo_rl/algorithms/loss_functions.py]: Sure. Done.

File: nemo_rl/algorithms/loss_functions.py [hunk 10]
```diff
@@ -73,12 +73,23 @@ class ClippedPGLossFn(LossFunction):
...
+    - c is the dual-clip parameter (ratio_clip_c), which must be greater than 1 and is
+      usually set as 3 empirically.
+
     Due to potential numerical instability, we cast the logits to float32 before computing the loss.
     """
 
     def __init__(self, cfg: ClippedPGLossConfig):
         self.ratio_clip_min = cfg["ratio_clip_min"]
         self.ratio_clip_max = cfg["ratio_clip_max"]
+        self.ratio_clip_c = cfg["ratio_clip_c"]
```
Comment (SahilJain314) [inline nemo_rl/algorithms/loss_functions.py]:
```suggestion
        self.ratio_clip_c = cfg["ratio_clip_c"] # set to None to disable dual-clipping
```
